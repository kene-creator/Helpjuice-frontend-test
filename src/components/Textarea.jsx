import { useState } from 'react';
import AddBlock from './AddBlock';

const Textarea = () => {
  const [text, setext] = useState('');
  const [openModal, setOpenModal] = useState(false);
  const [placeholder, setPlaceholder] = useState(true);
  const [header, setHeader] = useState('');
  const [enterPressed, setEnterPressed] = useState(false);

  const handleChange = (e) => {
    setext(e.target.value);
    if (e.target.value.includes('/1')) {
      return setOpenModal(true);
    }
    if (e.target.value.includes('') || e.target.value.includes('/')) {
      setOpenModal(false);
    }
  };

  const handleHeaderChange = (e) => {
    setHeader(e.target.value);
    if (e.target.value.includes('')) {
      setOpenModal(false);
    }
  };

  const handleKeyDown = (e) => {
    if (e.key === 'Enter') {
      setEnterPressed((prevState) => !prevState);
      setext('');
    }
  };

  return (
    <div>
      <div>
        {enterPressed && (
          <h1 className="text-[2.5rem] font-bold flex justify items-center w-[50%] mx-auto mt-4">
            {header}
          </h1>
        )}
      </div>
      {!enterPressed && (
        <textarea
          name=""
          id=""
          cols="30"
          rows="10"
          className="relative bg-white w-[50%] h-[50vh] mx-auto flex justify-center items-center mt-6 textarea_shadow placeholder:text-[1rem] text-lg"
          onChange={placeholder ? handleChange : handleHeaderChange}
          onKeyDown={handleKeyDown}
          placeholder={`${
            placeholder
              ? 'Type / for blocks, @ to link docs to people'
              : 'Heading 1'
          }`}
          value={!placeholder ? header : text}
        ></textarea>
      )}
      {enterPressed && (
        <textarea
          name=""
          id=""
          cols="30"
          rows="10"
          className="relative bg-white w-[50%] h-[50vh] mx-auto flex justify-center items-center mt-6 textarea_shadow placeholder:text-[1rem] text-lg"
          onChange={enterPressed ? handleChange : handleHeaderChange}
          onKeyDown={handleKeyDown}
          placeholder={`${
            enterPressed
              ? 'Now this is normal text. All I had to is press / +1, and then type my text and hit Enter/Return'
              : 'Heading 1'
          }`}
          value={enterPressed ? text : header}
        ></textarea>
      )}
      <div className="absolute top-[20rem] left-[15rem]">
        {openModal && <AddBlock callback={setPlaceholder} />}
      </div>
    </div>
  );
};

export default Textarea;
