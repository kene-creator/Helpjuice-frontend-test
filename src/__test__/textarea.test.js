import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Textarea from '../components/Textarea';

describe('Textarea', () => {
  it('should render textarea', () => {
    const { getByPlaceholderText } = render(<Textarea />);
    const textarea = getByPlaceholderText(
      'Type / for blocks, @ to link docs to people'
    );
    expect(textarea).toMatchSnapshot();
  });

  it('should open modal when /1 is typed', () => {
    const { getByPlaceholderText, getByText } = render(<Textarea />);
    const textarea = getByPlaceholderText(
      /Type \/ for blocks, @ to link docs to people/
    );

    fireEvent.change(textarea, { target: { value: '/1' } });
    expect(getByText('Add Block')).toMatchSnapshot();
  });

  it('should close modal when text is deleted', () => {
    const { getByPlaceholderText, queryByText } = render(<Textarea />);
    const textarea = getByPlaceholderText(
      /Type \/ for blocks, @ to link docs to people/
    );

    fireEvent.change(textarea, { target: { value: '/1' } });
    fireEvent.change(textarea, { target: { value: '' } });
    expect(queryByText('Add Block')).toBeNull();
  });

  it('should change placeholder when Enter is pressed', () => {
    const { getByPlaceholderText, queryByPlaceholderText } = render(
      <Textarea />
    );
    const textarea = getByPlaceholderText(
      /Type \/ for blocks, @ to link docs to people/
    );

    fireEvent.change(textarea, { target: { value: 'test' } });
    fireEvent.keyDown(textarea, { key: 'Enter', code: 13 });
    expect(
      queryByPlaceholderText(/Type \/ for blocks, @ to link docs to people/)
    ).toBeNull();
    expect(
      getByPlaceholderText(
        'Now this is normal text. All I had to is press / +1, and then type my text and hit Enter/Return'
      )
    ).toMatchSnapshot();
  });

  it('should change textarea value when Enter is pressed', () => {
    const { getByPlaceholderText, getByDisplayValue } = render(<Textarea />);
    const textarea = getByPlaceholderText(
      /Type \/ for blocks, @ to link docs to people/
    );

    fireEvent.change(textarea, { target: { value: 'test' } });
    fireEvent.keyDown(textarea, { key: 'Enter', code: 13 });
    expect(getByDisplayValue('')).toMatchSnapshot();
  });
});
