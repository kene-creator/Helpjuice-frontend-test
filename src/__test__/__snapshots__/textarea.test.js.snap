// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea should change placeholder when Enter is pressed 1`] = `
<textarea
  class="relative bg-white w-[50%] h-[50vh] mx-auto flex justify-center items-center mt-6 textarea_shadow placeholder:text-[1rem] text-lg"
  cols="30"
  id=""
  name=""
  placeholder="Now this is normal text. All I had to is press / +1, and then type my text and hit Enter/Return"
  rows="10"
/>
`;

exports[`Textarea should change textarea value when Enter is pressed 1`] = `
<textarea
  class="relative bg-white w-[50%] h-[50vh] mx-auto flex justify-center items-center mt-6 textarea_shadow placeholder:text-[1rem] text-lg"
  cols="30"
  id=""
  name=""
  placeholder="Now this is normal text. All I had to is press / +1, and then type my text and hit Enter/Return"
  rows="10"
/>
`;

exports[`Textarea should open modal when /1 is typed 1`] = `
<p
  class="font-bold text-[1rem]"
>
  Add Block
</p>
`;

exports[`Textarea should render textarea 1`] = `
<textarea
  class="relative bg-white w-[50%] h-[50vh] mx-auto flex justify-center items-center mt-6 textarea_shadow placeholder:text-[1rem] text-lg"
  cols="30"
  id=""
  name=""
  placeholder="Type / for blocks, @ to link docs to people"
  rows="10"
/>
`;
